### PR TITLE
allow developer to edit PayPal sdk query parameters

### LINKED ISSUE #4886: Epic: PayPal Commerce
# Pitch

This pitch has been defined in a [Google Doc](https://docs.google.com/document/d/1wD-Cyy4X6fDRe-tW4Ey3wSglcbYRAZQyDvjIEqVarRE/edit#heading=h.9g67mdj3rh9m).

The scope of this Epic is the integration of the new PayPal Commerce Platform payment gateway within GiveWP core.

## Note

This is the integration of the PayPal Commerce Platform API, but it will be called the **PayPal Checkout** gateway in GiveWP.

## Appetite

6 weeks, though we should finish as close to July 31 as possible

### LINKED ISSUE #4998: PayPal Commerce checkout sdk should extentable
## User Story

As a developer, I want to customize the PayPal SDK so that I can toggle any feature or payment method if I need.

## Details
PayPal SDK is customizable with query parameters: https://developer.paypal.com/docs/checkout/reference/customize-sdk/#query-parameters and currently, we load PayPal SDK with custom configuration which is not editable.

## Acceptance Criteria
-   [x] Disable `Debit or Credit Card` smart button as we already have an advanced card field on the form (Thank @JasonTheAdams for the suggestion ).
-   [x] We have a filter hook to edit PayPal SDK configuration.

### PR DESCRIPTION
**Pull request by ravinderk**
Resolves #4998 

## Description
PayPal SDK API provide various query parameter with which we can control the smart button and advanced card field output. I added a filter to edit the default SDK config. In addition to that I disabled `card` and `credit` smart button options.

I think I am using their option to disable payment options right away. I tested on there interactive demo and changes are not reflecting which means I am not able to disable the `card` option.
Slack discussion: https://givewp.slack.com/archives/C015XD31J5S/p1596751233004400

## Affects
- I updated code to do not load public assets if the user is connected to PayPal.

## Pre-review Checklist
-   [x] Acceptance criteria satisfied and marked in a related issue
-   [x] Relevant `@since` tags included in DocBlocks
-   [x] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed

## Testing Instructions
- Visit homepage when the user id does not connect to PayPal, you will not get any javascript error.
- You can also use filters to edit or add an available PayPal option. (link given in filter description)

### CODE REVIEW

**Comment by JasonTheAdams** [comment]
@ravinderk Is this supposed to resolve #4998? Right now it's set up to resolve the whole epic... and that's not right. 😅

**Comment by ravinderk** [comment]
@JasonTheAdams I attached PR to the correct the issue.

File: src/PaymentGateways/PayPalCommerce/ScriptLoader.php [hunk 1]
```diff
@@ -76,18 +78,34 @@ function givePayPalOnBoardedCallback(authCode, sharedId) {
 	 * @since 2.8.0
 	 */
 	public function loadPublicAssets() {
+		if ( ! Utils::isConnected() ) {
+			return;
+		}
```
Comment (JasonTheAdams) [inline src/PaymentGateways/PayPalCommerce/ScriptLoader.php:83]: This helper function doesn't exist anymore. You'll need to use the `MerchantDetails` repository.

File: src/PaymentGateways/PayPalCommerce/ScriptLoader.php [hunk 2]
```diff
@@ -29,27 +29,30 @@ class ScriptLoader {
...
-						'confirmPaypalAccountDisconnection' => esc_html__( 'Confirm PayPal account disconnection', 'give' ),
-						'disconnectPayPalAccount' => esc_html__( 'Do you want to disconnect PayPal account?', 'give' ),
-					],
-				]
-			);
-
-			$script = <<<EOT
+		if ( ! Utils::isConnected() || ! Give_Admin_Settings::is_setting_page( 'gateway', 'paypal' ) ) {
+			return;
+		}
```
Comment (JasonTheAdams) [inline src/PaymentGateways/PayPalCommerce/ScriptLoader.php:34]: `Utils::isConnected` no longer exists.

File: assets/src/js/frontend/paypal-commerce/SmartButtons.js [hunk 3]
```diff
@@ -36,6 +37,7 @@ class SmartButtons extends PaymentMethod {
 				color: 'gold',
 				tagline: false,
 			},
+			funding: { disallowed: [ paypal.FUNDING.CARD, paypal.FUNDING.CREDIT ] },
```
Comment (JasonTheAdams) [inline assets/src/js/frontend/paypal-commerce/SmartButtons.js:40]: We know that this doesn't work. It's also not really in scope with this PR. Let's remove this (and the comment above), for now.

**Comment by ravinderk** [comment]
@JasonTheAdams I think pr has old code. I will update it and let you know.

File: assets/src/js/frontend/paypal-commerce/SmartButtons.js (hunk 3 above)
Comment (ravinderk) [reply assets/src/js/frontend/paypal-commerce/SmartButtons.js]: fixed: https://github.com/impress-org/givewp/compare/896c0482ddb9...30b522638fc9

File: src/PaymentGateways/PayPalCommerce/ScriptLoader.php (hunk 1 above)
Comment (ravinderk) [reply src/PaymentGateways/PayPalCommerce/ScriptLoader.php]: fixed: https://github.com/impress-org/givewp/pull/5021/commits/dd79ab0fbe1020fe27e0df2f6dbc2c1a622307b7

File: src/PaymentGateways/PayPalCommerce/ScriptLoader.php (hunk 2 above)
Comment (ravinderk) [reply src/PaymentGateways/PayPalCommerce/ScriptLoader.php]: fixed: https://github.com/impress-org/givewp/pull/5021/commits/896c0482ddb97c2aa87c41f3d083075b8841b246

**Comment by ravinderk** [comment]
@JasonTheAdams this pr is ready to review.

**Comment by ravinderk** [comment]
@JasonTheAdams As we discussed, I removed the filter from this pr. This pr only contains refactored code which is safe to merge. After merging this pr, I will close the issue with a comment (about not adding filter).

**Pull request merged**